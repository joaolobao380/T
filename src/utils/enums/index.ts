export enum messageToast {
  ERROR = 'customError',
  SUCCESSFUL = 'successful',
}

export enum errorMessageFirebase {
  TOO_MANY = 'auth/too-many-requests',
  INVALID_PASSWORD = 'auth/invalid-password',
  INVALID_EMAIL = 'auth/invalid-email',
  INVALID_CREDENTIAL = 'auth/invalid-credential',
}

export enum textToast {
  ATTENTION = 'Atenção',
  SUCCESSFUL = 'Sucesso!',
}

export enum positionToast {
  POSITION = 'bottom',
}

export enum keyAsyncUser {
  USER_TOKEN_KEY = '@userToken',
}

export enum navBottomName {
  HOME = 'Home',
  REPORT = 'Report',
  MENU = 'Menu',
}
