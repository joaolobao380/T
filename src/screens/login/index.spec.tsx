// CustomInput.test.tsx
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import { messageToast, errorMessageFirebase } from '@utils/enums';
import React from 'react';
import Toast from 'react-native-toast-message';

import { Login } from './index';

const emailTest = 'test@example.com';
const placeholderEmail = 'Ex: email@email.com';
const placeholderPassword = 'Ex: •••••••••••';
const nameFieldPassword = 'password';
const userUuid = 'uuid do usuário';
const titleButton = 'Entrar';

const mockNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
  };
});

jest.mock('firebase/app', () => {
  return {
    initializeApp: jest.fn(() => {}),
  };
});

jest.mock('firebase/auth', () => {
  return {
    getAuth: jest.fn(() => ({})),
    signInWithEmailAndPassword: jest.fn(() =>
      Promise.resolve({
        user: { email: emailTest, uuid: userUuid },
      })
    ),
    signOut: jest.fn(() => Promise.resolve()),
    onAuthStateChanged: jest.fn((auth, callback) => {
      callback({ email: emailTest, uuid: userUuid });
      return jest.fn();
    }),
    sendPasswordResetEmail: jest.fn((auth, email) => Promise.resolve()),
  };
});

jest.mock('react-native-toast-message', () => {
  return {
    show: jest.fn(),
  };
});

const mockLogin = jest.fn();
jest.mock('@hooks/useAuth', () => ({
  __esModule: true,
  default: () => ({
    login: mockLogin,
  }),
}));

describe('Login screen', () => {
  it('Should call login function with correct credentials on submit', async () => {
    const { getByPlaceholderText, getByText } = render(<Login />);
    const emailInput = getByPlaceholderText(placeholderEmail);
    const passwordInput = getByPlaceholderText(placeholderPassword);
    const submitButton = getByText(titleButton);

    fireEvent.changeText(emailInput, emailTest);
    fireEvent.changeText(passwordInput, nameFieldPassword);
    fireEvent.press(submitButton);

    await waitFor(() => expect(mockLogin).toHaveBeenCalledWith(emailTest, nameFieldPassword));
  });

  it('Should show error toast on login failure auth/invalid-credential', async () => {
    mockLogin.mockRejectedValueOnce({ code: errorMessageFirebase.INVALID_CREDENTIAL });

    const { getByPlaceholderText, getByText } = render(<Login />);
    const emailInput = getByPlaceholderText(placeholderEmail);
    const passwordInput = getByPlaceholderText(placeholderPassword);
    const submitButton = getByText(titleButton);

    fireEvent.changeText(emailInput, emailTest);
    fireEvent.changeText(passwordInput, nameFieldPassword);
    fireEvent.press(submitButton);

    await waitFor(
      () => {
        expect(Toast.show).toHaveBeenCalledWith(
          expect.objectContaining({
            type: messageToast.ERROR,
            text2: 'Senha incorreta. Por favor, tente novamente.',
          })
        );
      },
      { timeout: 1000 }
    );
  });

  it('Should show error toast on login failure auth/too-many-requests', async () => {
    mockLogin.mockRejectedValueOnce({ code: errorMessageFirebase.TOO_MANY });

    const { getByPlaceholderText, getByText } = render(<Login />);
    const emailInput = getByPlaceholderText(placeholderEmail);
    const passwordInput = getByPlaceholderText(placeholderPassword);
    const submitButton = getByText(titleButton);

    fireEvent.changeText(emailInput, emailTest);
    fireEvent.changeText(passwordInput, nameFieldPassword);
    fireEvent.press(submitButton);

    await waitFor(
      () => {
        expect(Toast.show).toHaveBeenCalledWith(
          expect.objectContaining({
            type: messageToast.ERROR,
            text2: 'Seu acesso foi bloqueado.',
          })
        );
      },
      { timeout: 1000 }
    );
  });

  it('Should show error toast on login failure auth/too-many-requests', async () => {
    mockLogin.mockRejectedValueOnce({ code: 'qualquer error' });

    const { getByPlaceholderText, getByText } = render(<Login />);
    const emailInput = getByPlaceholderText(placeholderEmail);
    const passwordInput = getByPlaceholderText(placeholderPassword);
    const submitButton = getByText(titleButton);

    fireEvent.changeText(emailInput, emailTest);
    fireEvent.changeText(passwordInput, nameFieldPassword);
    fireEvent.press(submitButton);

    await waitFor(
      () => {
        expect(Toast.show).toHaveBeenCalledWith(
          expect.objectContaining({
            type: messageToast.ERROR,
            text2: 'Não foi possível logar-se, tente novamente.',
          })
        );
      },
      { timeout: 1000 }
    );
  });

  it('Should navigate to ForgotPassword screen on forgot password button press', () => {
    const { getByText } = render(<Login />);
    const forgotPasswordButton = getByText('Esqueci minha senha');
    fireEvent.press(forgotPasswordButton);
    expect(mockNavigate).toHaveBeenCalledWith('ForgotPassword');
  });

  it('renders correctly', () => {
    const { toJSON } = render(<Login />);
    expect(toJSON()).toMatchSnapshot();
  });
});
