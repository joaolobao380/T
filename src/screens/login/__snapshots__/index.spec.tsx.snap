// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "#2A2A2A",
        "flex": 1,
        "flexGrow": 1,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView
    collapsable={false}
    handlerTag={6}
    handlerType="NativeViewGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    style={
      {
        "backgroundColor": "white",
        "flex": 1,
      }
    }
    waitFor={
      [
        {
          "current": null,
        },
      ]
    }
  >
    <View>
      <View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Image
            source={1}
            style={
              {
                "marginBottom": 16,
                "width": "100%",
              }
            }
          />
        </View>
        <View
          style={
            {
              "marginTop": 40,
              "paddingHorizontal": 16,
            }
          }
        >
          <View>
            <View
              style={
                {
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#2A2A2A",
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 14,
                    "marginBottom": 4,
                  }
                }
              >
                E-mail
              </Text>
              <TextInput
                keyboardType="email-address"
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Ex: email@email.com"
                placeholderTextColor="#BBBBBB"
                secureTextEntry={false}
                style={
                  [
                    {
                      "backgroundColor": "white",
                      "borderColor": "#2A2A2A",
                      "borderRadius": 5,
                      "borderWidth": 1,
                      "color": "#2A2A2A",
                      "fontFamily": "Poppins-Regular",
                      "fontSize": 16,
                      "padding": 16,
                    },
                    false,
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              {
                "marginTop": 16,
              }
            }
          >
            <View
              style={
                {
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#2A2A2A",
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 14,
                    "marginBottom": 4,
                  }
                }
              >
                Senha
              </Text>
              <TextInput
                keyboardType="default"
                onBlur={[Function]}
                onChangeText={[Function]}
                placeholder="Ex: •••••••••••"
                placeholderTextColor="#BBBBBB"
                secureTextEntry={true}
                style={
                  [
                    {
                      "backgroundColor": "white",
                      "borderColor": "#2A2A2A",
                      "borderRadius": 5,
                      "borderWidth": 1,
                      "color": "#2A2A2A",
                      "fontFamily": "Poppins-Regular",
                      "fontSize": 16,
                      "padding": 16,
                    },
                    false,
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              {
                "marginTop": 32,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#2A2A2A",
                    "borderRadius": 4,
                    "justifyContent": "center",
                    "paddingVertical": 16,
                    "width": "100%",
                  },
                  false,
                  false,
                ]
              }
            >
              <Text
                style={
                  {
                    "color": "#FFFFFF",
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 16,
                  }
                }
              >
                Entrar
              </Text>
            </View>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "marginTop": 24,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#2A2A2A",
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 16,
                  }
                }
              >
                Esqueci minha senha
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
